Add option to use SNOPT's internal finite difference method

jaredthomas68 opened the pull request.

Kenneth-T-Moore commented on openmdao/drivers/pyoptsparse_driver.py:
```diff
@@ -276,6 +280,13 @@ def run(self, problem):
+                sol = opt(opt_prob, sens=None, sensStep=fd_step, storeHistory=self.hist_file)
+            else:
+                msg = "SNOPT's internal finite difference method can only be used with SNOPT"
+                raise Exception(msg)
```
The only thing I'd add is one more test where we set snopt_diff to True, but use one of the other optimizers so that we get test coverage on the error message. Aside from that small thing, this pull request looks pretty solid.

naylor-b replied:
We now have 3 mutually exclusive possibilities in terms of selecting the gradient method.  I suggest that instead of having two different flags (pyopt_diff and snopt_diff), We use a single enum, maybe 'gradient_method', with 3 values, maybe 'pyopt_fd', 'snopt_fd', and 'openmdao', with 'openmdao' being the default.  The 'add_option' command has a 'values' arg that lets you set enumerated values.  If we keep two separate flags, we would need to add error checking to make sure that both flags were not set at the same time.  Currently if pyopt_diff is set, snopt_diff is ignored.

jaredthomas68 replied:
That is a good point, I agree.

naylor-b replied:
I spoke with @JustinSGray about this and he agrees that we should change the option as I described.  Can you incorporate that change into your PR?

jaredthomas68 replied:
Yes, I'll do that.

Kenneth-T-Moore commented on openmdao/drivers/test/test_pyoptsparse.py:
```diff
@@ -1191,5 +1191,38 @@ def solve_nonlinear(self, params, unknowns, resids):
+        assert_rel_error(self, prob['x'], 7.16667, 1e-6)
+        assert_rel_error(self, prob['y'], -7.833334, 1e-6)
+
+        
```
One thing you could do to be absolutely certain that the openmdao derivatives aren't still being called (and to test it) is to create a new class that inherits from `Paraboloid`, and then give it an `apply_linear` function that does nothing except raise an exception if called. The apply_linear function would get called if openmdao took a derivative.

jaredthomas68 commented:
@Kenneth-T-Moore, thank you for your suggestions, both have been addressed in my last commit.

jaredthomas68 commented:
appveyor: I can't see where this issue relates to my changes. Ideas?

travis-cl: The only failing test here that I have touched was the test of pyopt_fd that I added. It passes locally but does not quite get to the required tolerance with travis-cl.

Coverage: I'm not sure why the coverage is showing a large decrease. When I run coverage locally I get 98% coverage for pyoptsparse_driver.py, which is the only thing I have changed.

Please let me know if you know what I can do to pass the checks.

kmarsteller commented:
You're OK on coverage.  That's the coverage of the first machine that finishes.  The last machine runs more tests and will report actual coverage.

kmarsteller commented:
You're OK on Appveyor, too.  This is an ongoing issue for us, not your fault.

kmarsteller commented:
The coverage won't be correct until the machine on Travis that runs all of the MPI tests (machine #1) actually finishes/passes.

naylor-b commented:
@jaredthomas68, if you update the tolerance on that one test so that it passes on travis I'll merge it.
